src/context/values.js

```diff
@@ -0,0 +1,720 @@
+          if (used.counter > 5) {
+            // A simple protection from infinte loops.
+            rethrowAsync(new Error('cyclical prop: ' + key));
+            used.pending = Pending.NOT_PENDING;
+            return;
+          }
+          updated++;
+          this.tryUpdate_(used);
+        }
+      });
+    } while (updated > 0);
+  }
+
+  /**
+   * @param {!UsedDef} used
+   * @private
+   */
+  tryUpdate_(used) {
+    // The value is not pending anymore. If any of the dependencies will remain
+    // unresolved, we will simply need to recomputed it.
+    const refreshParent = used.pending == Pending.PENDING_REFRESH_PARENT;
+
+    let newValue;
+    try {
+      newValue = this.calc_(used, refreshParent);
+    } catch (e) {
+      // This is the narowest catch to avoid unrelated values breaking each
+      // other. The only exposure to the user-code are `needsParent` and
+      // `compute` methods in the `ContextProp`.
+      rethrowAsync(e);
+    }
+
+    // Reset pending flag. It's good to reset it after the calculation to
+    // ensure that deps are automatically covered.
+    used.pending = Pending.NOT_PENDING;
+
+    // Check if the value has been updated.
+    this.maybeUpdated_(used, newValue);
+  }
+
+  /**
+   * @param {!UsedDef} used
+   * @param {*} value
+   * @private
+   */
+  maybeUpdated_(used, value) {
+    const {prop, value: oldValue} = used;
+    const {key} = prop;
+    const usedByKey = this.usedByKey_;
+    if (
+      oldValue === value ||
+      used !== (usedByKey && usedByKey.get(key)) ||
+      !this.isConnected_()
+    ) {
+      // Either the value didn't change, or no one needs this value anymore.
+      return;
+    }
+
+    used.value = value;
+
+    // Notify subscribers.
+    const {subscribers} = used;
+    subscribers.forEach((handler) => {
+      handler(value);
+    });
+  }
+
+  /**
+   * The used value calculation algorithm.
+   *
+   * @param {!UsedDef} used
+   * @param {boolean} refreshParent
+   * @return {*|undefined} The used value.
+   * @private
+   */
+  calc_(used, refreshParent) {
+    devAssert(this.isConnected_());
+
+    const {prop, depValues} = used;
+    const {key, compute, defaultValue} = prop;
+
+    const inputsByKey = this.inputsByKey_;
+    const inputs = inputsByKey && inputsByKey.get(key);
+    const inputValues = inputs && inputs.values;
+
+    // Calculate parent value.
+    const needsParent = calcNeedsParent(prop, inputValues);
+
+    // Refresh parent if requested.
+    if (refreshParent || needsParent != Boolean(used.parentValueNode)) {
+      const newParentValueNode = needsParent
+        ? findParent(this.contextNode_, hasInput, prop, /* includeSelf */ false)
+        : null;
+      this.updateParentValueNode_(used, newParentValueNode);
+    }
+
+    // If no parent node is found, use the default value.
+    const parentValue = defined(used.parentValue)
+      ? used.parentValue
+      : needsParent && !used.parentValueNode && inputValues
```

Comment:
Why is `inputValues` required to resolve to the `defaultValue`?

Reply:
Hmm. `inputValues` do not resolve to the `defaultValue`. Basically, this condition evaluates as:

1. The `parentValue` is required (b/c inputs are available and parent is requested).
2. The `defaultValue` is specified.

In this case, the code falls back to the `defaultValue` until a new parent value is available.

Reply:
I think I misunderstood your question a bit. Let me try to answer this again: there are two reasons why a parent value might be needed:
 1. The prop value is requested on a node, but the node itself doesn't have any inputs. So the value is simply taken from the parent. However, if the parent doesn't have a value - nothing to take any way.
 2. The node has inputs and the prop value will be computed as `usedValue = compute(inputs, parentValue, ...)`. In this case, we will pass the `defaultValue` for parent if it's not available.

This is similar to CSS. E.g. if I were to specify `font-size: 80%`, I really need to know the parent's `font-size` (or default) to compute the `80%` from it.

Reply:
I'm also confused why we need `inputValues` here… It doesn't seem to be related to using a `defaultValue` at all. If we need a parent, and I don't have one, then default. If I don't need a parent, then `undefined`.

Reply:
Based on my understanding of the first two descriptions, would it be accurate to give these variable names?
```js
const isParentValueRequired = needsParent && !used.parentValueNode && inputValues;
const placeholderForParentValue = defaultValue;
```

If this were the case, then it seems like the logic of this line is:
`isParentValueDefined ? parentValue : isParentValueRequired ? placeholderForParentValue : undefined` -- this seems to align with the inline comment `// If no parent node is found, use the default value.`

Follow up questions:
- Does `needsParent` imply `inputValues != null`?
- How we would know this is a sensible default, or whether or not we should need to?
- If it is appropriate to call this `defaultValue` a placeholder until the parent value is resolved, at what point does it get replaced?

Reply:
I think you're both right. I removed `inputValues` from that calculation. There's some use case still there to cover to a default value that would only be needed for `compute` function. But I think it's separate enough case and I can come back to it later. Thus, the `parentValue` is now simplified per your asks.